neqo-transport/src/cc/tests/cubic.rs

```diff
@@ -0,0 +1,303 @@
+        next_pn_send = fill_cwnd(&mut cubic, next_pn_send, now);
+    }
+
+    let expected_ack_tcp_increase2 = expected_tcp_acks(cwnd_rtt_start_after_tcp);
+    assert!(num_acks2 < expected_ack_tcp_increase2);
+
+    // The time needed to increase cwnd by MAX_DATAGRAM_SIZE using the cubic equation will be calculates from:
+    // W_cubic(elapsed_time + t_to_increase) - W_cubis(elapsed_time) = MAX_DATAGRAM_SIZE =>
+    // CUBIC_C * (elapsed_time + t_to_increase)^3 * MAX_DATAGRAM_SIZE + CWND_INITIAL -
+    //     CUBIC_C * elapsed_time^3 * MAX_DATAGRAM_SIZE + CWND_INITIAL = MAX_DATAGRAM_SIZE =>
+    // t_to_increase = cbrt((1 + CUBIC_C * elapsed_time^3) / CUBIC_C) - elapsed_time
+    // (t_to_increase is in seconds)
+    // number of ack needed is t_to_increase / time_increase.
+    let expected_ack_cubic_increase =
+        ((((1.0 + CUBIC_C * (elapsed_time).as_secs_f64().powi(3)) / CUBIC_C).cbrt()
+            - elapsed_time.as_secs_f64())
+            / time_increase.as_secs_f64())
+        .ceil() as u64;
+    // num_acks is very close to the calculated value. The exact value is hard to calculate
+    // because the proportional increase(i.e. curr_cwnd_f64 / (target - curr_cwnd_f64) * MAX_DATAGRAM_SIZE_F64)
+    // and the byte counting.
+    assert_eq!(num_acks2, expected_ack_cubic_increase + 2);
+}
+
+#[test]
+fn cubic_phase() {
+    let mut cubic = ClassicCongestionControl::new(Cubic::default());
+    // Set last_max_cwnd to a higher number make sure that cc is the cubic phase (cwnd is calculated by the cubic equation).
+    cubic.set_last_max_cwnd(CWND_INITIAL_10_F64);
+    // Set ssthresh to something small to make sure that cc is in the congection avoidance phase.
+    cubic.set_ssthresh(1);
+    let mut now = now();
+    let mut next_pn_send = 0;
+    let mut next_pn_ack = 0;
+
+    next_pn_send = fill_cwnd(&mut cubic, next_pn_send, now);
+
+    let k = ((CWND_INITIAL_10_F64 - CWND_INITIAL_F64) / CUBIC_C / MAX_DATAGRAM_SIZE_F64).cbrt();
+    let epoch_start = now;
+
+    // The number of RTT until W_max is reached.
+    let num_rtts_w_max = (k / RTT.as_secs_f64()).round() as u64;
+    for _ in 0..num_rtts_w_max {
+        let cwnd_rtt_start = cubic.cwnd();
+        //Expected acks
+        let acks = cwnd_rtt_start / MAX_DATAGRAM_SIZE;
+        let time_increase = RTT / u32::try_from(acks).unwrap();
+        for _ in 0..acks {
+            now += time_increase;
+            ack_packet(&mut cubic, next_pn_ack, now);
+            next_pn_ack += 1;
+            next_pn_send = fill_cwnd(&mut cubic, next_pn_send, now);
+        }
+
+        let expected =
+            (CUBIC_C * ((now - epoch_start).as_secs_f64() - k).powi(3) * MAX_DATAGRAM_SIZE_F64
+                + CWND_INITIAL_10_F64)
+                .round() as usize;
+
+        assert_within(cubic.cwnd(), expected, &MAX_DATAGRAM_SIZE);
+    }
+    assert_eq!(cubic.cwnd(), CWND_INITIAL_10);
+}
+
+fn assert_within<T: Sub<Output = T> + PartialOrd>(value: T, expected: T, margin: &T) {
```

Comment:
You can make `T: Sub<...> + PartialOrd + Copy` and avoid taking a reference for `margin`.